In [1]:
import pandas as pd
import numpy as np
import os
import glob
from pandas import DataFrame, read_csv
import matplotlib.pyplot as plt
import pandas as pd
import wfdb
import statsmodels.api as sm
import matplotlib.image as mpimg
import tkinter as tk
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
from matplotlib.backends.backend_tkagg import NavigationToolbar2Tk
from matplotlib.figure import Figure
from IPython.display import display, HTML
import csv
from collections import defaultdict
import matplotlib.ticker as ticker
import pickle
import matplotlib as mpl
from sklearn.metrics import r2_score
import warnings
import importlib
from RDFunctions import *

In [4]:
annofile = r'RecordsAnno3.csv'
dfOuranno = pd.read_csv(annofile, encoding='iso-8859-1') 
d = ['SegmentId','RecordNum','StartingTime','RRBreach','SpO2Breach','RRHighRegions','SpO2LowRegions','Stage1OccurL','Stage2OccurL','Stage3OccurL','RRUnsafeRegions','SpO2UnsafeRegions','Stage1OccurH','Stage2OccurH','Stage3OccurH','FeatureSum','Type1Pattern']
segFeatureFrame = pd.DataFrame(columns = d)
RECORDS = np.shape(dfOuranno)[0]
for i in range(RECORDS):
    if ((dfOuranno.iloc[i]['Category']=='A' or dfOuranno.iloc[i]['Category']=='E') and pd.notnull(dfOuranno.iloc[i]['DTW_RDSegs'])): 
        print(i)
        mystr = dfOuranno.iloc[i]['DTW_RDSegs'].split(';')
        L = len(mystr)
        for j in range(L-1):
            start = float(mystr[j].split(':')[0])
            try:
                addFeatureVectorToSegFrame(i,dfOuranno,start,1,segFeatureFrame)
            except ValueError:
                break
for i in range(RECORDS):
    if ((dfOuranno.iloc[i]['Category']=='C' or dfOuranno.iloc[i]['Category']=='F') and pd.notnull(dfOuranno.iloc[i]['DTW_RDSegs'])): 
        print(i)
        mystr = dfOuranno.iloc[i]['DTW_RDSegs'].split(';')
        L = len(mystr)
        for j in range(L-1):
            start = float(mystr[j].split(':')[0])
            try:
                addFeatureVectorToSegFrame(i,dfOuranno,start,0,segFeatureFrame)
            except ValueError:
                break

13
19
23
25
39
43
45
46
50
54
66
67
69
70
78
82
86
87
91
93
114
116
117
122
128
130
132
139
151
158
160
162
166
168
169
176
178
181
183
185
188
200
206
207
209
212
213
218
227
229
232
238
248
253
267
270
275
299
300
320
323
328
333
342
344
346
352
354
356
360
362
365
367
368
373
376
379
382
383
385
386
389
391
397
403
404
412
415
417
423
428
432
436
437
442
450
455
462
464
469
475
476
490
492
493
508
513
515
520
523
524
545
548
554
555
556
557
562
563
566
568
571
575
585
598
601
610
612
613
614
620
623
630
645
650
651
653
654
662
670
677
679
680
697
701
706
708
712
720
723
734
735
737
744
750
752
758
760
766
767
772
779
785
790
791
794
803
812
815
816
818
823
824
825
826
836
837
838
839
845
857
859
860
867
870
885
887
890
893
894
900
904
908
909
2
3
4
5
8
11
16
18
20
21
29
34
35
36
48
52
64
74
76
88
89
90
95
96
98
102
103
105
106
112
113
120
123
125
126
127
133
136
138
140
146
149
150
152
155
156
157
159
161
167
170
171
172
179
182
189
198
199
204
208
210
222
228
231
234
236
237
241
24

In [5]:
display(segFeatureFrame)

SegmentId RecordNum  StartingTime RRBreach SpO2Breach RRHighRegions  \
1         Seg0   3861446          30.0        0          0             0   
2         Seg1   3861446          31.0        0          0             0   
3         Seg2   3861446          32.0        0          0             0   
4         Seg3   3861446           4.0        0          1             0   
5         Seg4   3861446           8.0        0          1             0   
6         Seg5   3861446           5.0        0          1             0   
7         Seg6   3861446           7.0        0          1             0   
8         Seg7   3861446           6.0        0          1             0   
9         Seg8   3060876          42.0        0          1             0   
10        Seg9   3060876          38.0        0          1             0   
11       Seg10   3060876          41.0        0          1             0   
12       Seg11   3060876          39.0        0          1             0   
13       Seg12   3060876          40.0        0          1             0   
14       Seg13   3976294          76.0        1          0            17   
15       Seg14   3976294          77.0        1          0            19   
16       Seg15   3976294          78.0        1          0            17   
17       Seg16   3976294           2.0        1          0            23   
18       Seg17   3976294           5.0        1          0            24   
19       Seg18   3976294           3.0        1          0            23   
20       Seg19   3976294           4.0        1          0            24   
21       Seg20   3653853          70.0        1          0             1   
22       Seg21   3653853          66.0        1          0             1   
23       Seg22   3653853          69.0        1          0             1   
24       Seg23   3653853          67.0        1          0             1   
25       Seg24   3653853          68.0        1          1             1   
26       Seg25   3414279          31.0        1          1             7   
27       Seg26   3414279          28.0        1          0             8   
28       Seg27   3414279          30.0        1          1             8   
29       Seg28   3414279          29.0        1          0             8   
30       Seg29   3818257          19.0        0          1             0   
...        ...       ...           ...      ...        ...           ...   
2709   Seg2708   3535257          32.0        0          0             0   
2710   Seg2709   3772248           0.0        0          0             0   
2711   Seg2710   3772248           2.0        0          0             0   
2712   Seg2711   3772248           4.0        0          0             0   
2713   Seg2712   3772248           6.0        0          0             0   
2714   Seg2713   3772248           8.0        0          0             0   
2715   Seg2714   3772248          10.0        0          0             0   
2716   Seg2715   3844715           0.0        0          1             0   
2717   Seg2716   3844715           2.0        0          1             0   
2718   Seg2717   3844715           4.0        0          0             0   
2719   Seg2718   3844715           6.0        0          0             0   
2720   Seg2719   3844715           8.0        0          0             0   
2721   Seg2720   3844715          10.0        0          0             0   
2722   Seg2721   3844715          12.0        0          0             0   
2723   Seg2722   3844715          14.0        0          0             0   
2724   Seg2723   3760745           0.0        0          0             0   
2725   Seg2724   3760745           2.0        0          0             0   
2726   Seg2725   3760745           4.0        0          0             0   
2727   Seg2726   3760745           6.0        0          0             0   
2728   Seg2727   3760745           8.0        0          0             0   
2729   Seg2728   3760745          10.0        0      

#### Ignore previous one because RecordsAnno4.csv is the one we are using now. It has 752 negative and 758 postive segments. Finally we create SegmentFeatures2.csv for ML models

In [6]:
segFeatureFrame.to_csv('SegmentFeatures.csv',index=False)

In [2]:
annofile = r'RecordsAnno4.csv'
dfOuranno = pd.read_csv(annofile, encoding='iso-8859-1') 
d = ['SegmentId','RecordNum','StartingTime','RRBreach','SpO2Breach','RRHighRegions','SpO2LowRegions','Stage1OccurL','Stage2OccurL','Stage3OccurL','RRUnsafeRegions','SpO2UnsafeRegions','Stage1OccurH','Stage2OccurH','Stage3OccurH','FeatureSum','Type1Pattern']
segFeatureFrame = pd.DataFrame(columns = d)
RECORDS = np.shape(dfOuranno)[0]
for i in range(RECORDS):
    if ((dfOuranno.iloc[i]['Category']=='A' or dfOuranno.iloc[i]['Category']=='E') and pd.notnull(dfOuranno.iloc[i]['DTW_RDSegs'])): 
        print(i)
        mystr = dfOuranno.iloc[i]['DTW_RDSegs'].split(';')
        L = len(mystr)
        for j in range(L-1):
            start = float(mystr[j].split(':')[0])
            try:
                addFeatureVectorToSegFrame(i,dfOuranno,start,1,segFeatureFrame)
            except ValueError:
                break
for i in range(RECORDS):
    if ((dfOuranno.iloc[i]['Category']=='C' or dfOuranno.iloc[i]['Category']=='F') and pd.notnull(dfOuranno.iloc[i]['DTW_RDSegs'])): 
        print(i)
        mystr = dfOuranno.iloc[i]['DTW_RDSegs'].split(';')
        L = len(mystr)
        for j in range(L-1):
            start = float(mystr[j].split(':')[0])
            try:
                addFeatureVectorToSegFrame(i,dfOuranno,start,0,segFeatureFrame)
            except ValueError:
                break

13
19
23
25
39
43
45
46
50
54
66
67
69
70
78
82
86
87
91
93
114
116
117
122
128
130
132
139
151
158
160
162
166
168
169
176
178
181
183
185
188
200
206
207
209
212
213
218
227
229
232
238
248
253
267
270
275
299
300
320
323
328
333
342
344
346
352
354
356
360
362
365
367
368
373
376
379
382
383
385
386
389
391
397
403
404
412
415
417
423
428
432
436
437
442
450
455
462
464
469
475
476
490
492
493
508
513
515
520
523
524
545
548
554
555
556
557
562
563
566
568
571
575
585
598
601
610
612
613
614
620
623
630
645
650
651
653
654
662
670
677
679
680
697
701
706
708
712
720
723
734
735
737
744
750
752
758
760
766
767
772
779
785
790
791
794
803
812
815
816
818
823
824
825
826
836
837
838
839
845
857
859
860
867
870
885
887
890
893
894
900
904
908
909
2
3
4
5
8
11
16
18
20
21
29
34
35
36
48
52
64
74
76
88
89
90
95
96
98
102
103
105
106
112
113
120
123
125
126
127
133
136
138
140
146
149
150
152
155
156
157
159
161
167
170
171
172
179
182
189
198
199
204
208
210
222
228
231
234
236
237
241
24

In [3]:
segFeatureFrame.to_csv('SegmentFeatures2.csv',index=False)